### PR TITLE
Get confirmed + unconfirmed unspent boxes EP

### PR DESCRIPTION
**Pull request by semyonoskin**
#245

### CODE REVIEW

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/services/Boxes.scala [hunk 1]
```diff
@@ -379,6 +391,28 @@ object Boxes {
         .thrushK(trans.xa)
     }
 
+    def getAllUnspentOutputs(address: Address, paging: Paging, ord: SortOrder): F[Items[UOutputInfo]] = {
+      val ergoTree = addressToErgoTreeHex(address)
+      (for {
+        nConfirmed  <- outputs.countUnspentByErgoTree(ergoTree)
+        nUnonfirmed <- uoutputs.sumUnspentByErgoTree(ergoTree)
```
Comment (oskin1) [inline modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/services/Boxes.scala:398]:
```suggestion
        nUnсonfirmed <- uoutputs.sumUnspentByErgoTree(ergoTree)
```

Comment (oskin1) [inline modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/services/Boxes.scala:398]: Please look careful at what `sumUnspentByErgoTree` really does
https://github.com/ergoplatform/explorer-backend/blob/44c1acbc3d0347eec17b2c2e5216ae5e23181d78/modules/explorer-core/src/main/scala/org/ergoplatform/explorer/db/queries/UOutputQuerySet.scala#L191

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/services/Boxes.scala [hunk 2]
```diff
@@ -379,6 +391,28 @@ object Boxes {
...
+                       .streamUnspentByErgoTree(ergoTree, paging.offset, paging.limit, ord.value)
+                       .chunkN(serviceSettings.chunkSize)
+                       .through(toOutputInfo)
+                       .map(UOutputInfo.fromOutputInfo)
+                       .to[List]
+        unconfirmed <- Stream
+                         .evalSeq(
+                           uoutputs
+                             .getAllUnspentByErgoTree(ergoTree)
+                         )
```
Comment (oskin1) [inline modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/services/Boxes.scala:409]: So it seems your paging doesn't work as it should here. You retrieve a slice of expected number of confirmed outputs `streamUnspentByErgoTree(ergoTree, paging.offset, paging.limit, ord.value)` and then concatenate it with unknown number of unconfirmed outputs.

**Review by oskin1 — CHANGES_REQUESTED** [review]

File: modules/explorer-core/src/main/scala/org/ergoplatform/explorer/db/queries/UOutputQuerySet.scala [hunk 3]
```diff
@@ -184,6 +185,49 @@ object UOutputQuerySet extends QuerySet {
...
+         |  o.additional_registers,
+         |  null
+         |from node_u_outputs o
+         |left join node_u_inputs i on i.box_id = o.box_id
+         |where i.box_id is null and o.ergo_tree = $ergoTree
+         |union all
+         |select distinct on (o.box_id, o.global_index)
+         |  o.box_id,
+         |  o.tx_id,
+         |  o.value,
```
Comment (oskin1) [inline modules/explorer-core/src/main/scala/org/ergoplatform/explorer/db/queries/UOutputQuerySet.scala:211]: Please indent SQL

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/defs/BoxesEndpointDefs.scala [hunk 4]
```diff
@@ -170,4 +170,11 @@ final class BoxesEndpointDefs[F[_]](settings: RequestsSettings) {
         "Search among UTXO set by ergoTreeTemplateHash and tokens. " +
         "The resulted UTXOs will contain at lest one of the given tokens."
       )
+
+  def getAllUnspentOutputsByAddressDef: Endpoint[(Address, Paging, SortOrder), ApiErr, Items[UOutputInfo], Any] =
+    baseEndpointDef.get
+      .in(PathPrefix / "unspent" / "all" / "byAddress" / path[Address])
+      .in(paging(settings.maxEntitiesPerRequest))
+      .in(ordering)
+      .out(jsonBody[Items[UOutputInfo]])
```
Comment (oskin1) [inline modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/defs/BoxesEndpointDefs.scala:179]: Let's not shrink `OutputInfo` down to `UOutputInfo` (omitting some fields). I'd suggest adding new coproduct that contains both variants (`AnyOutputInfo = OutputInfo | UOutputInfo`).

**Review by oskin1 — APPROVED** [review]

**Comment by GusevTimofey** [comment]
https://github.com/ergoplatform/explorer-backend/blob/master/modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/defs/BoxesEndpointDefs.scala#L133

may help you and can be used as reference

**Pull request merged**